feat(ledc): max resolution review

SuGlider opened the pull request:
## Description of Change
LEDC says that maximum duty value is 2^resolution_bits
Arduino `analogWrite()` says that the maximum duty is (2^resolution_bits) - 1
Example for 8 bits resolution, LEDC max duty is 256 and Arduino `analogWrite()` is 255.

This PR changes the code to reflec exactly that. 
When using LEDC functions, it will allow max duty to go to 2^resolution_bits.
When using Arduino API `analogWrite()`, it will map duty (2^resolution_bits) - 1 to maximum value in LEDC before calling its function.

By this way both are compatible to its own specification and allows all valid values to be used.

## Tests scenarios
ESP32 and ESP32C6 using digital Logic Analyzer.

## Related links
Fix #11049

SuGlider committed:
feat(ledc): max resolution review

SuGlider committed:
feat(ledc): check duty limit based on the resolution

SuGlider commented on cores/esp32/esp32-hal-ledc.c:
```diff
@@ -183,14 +183,13 @@ bool ledcWrite(uint8_t pin, uint32_t duty) {
+      duty = max_duty;
     }
-
+ 
```
```suggestion

```

SuGlider committed:
fix(ledc): removing white space in code line

SuGlider committed:
fix(ledc): reverting not necessary change

SuGlider committed:
feat(ledc): testing values that shall be consistent

SuGlider commented on cores/esp32/esp32-hal-ledc.c:
```diff
@@ -446,6 +450,17 @@ void analogWrite(uint8_t pin, int value) {
+    // Therefore, it will adjust Arduino MAX Duty to be the LEDC MAx Duty
+    uint32_t max_duty = (1 << bus->channel_resolution) - 1;
+    if (value < 0 || value > max_duty) {
+      log_w("Duty is out of range. Valid duty range for pin d is 0 to %d", pin, max_duty);
```
```suggestion
      log_we"Duty is out of range. Valid duty range for pin d is 0 to %d", pin, max_duty);
```

SuGlider committed:
fix(ledc): failing to execute shall issue error log

SuGlider commented on cores/esp32/esp32-hal-ledc.c:
```diff
@@ -446,6 +450,17 @@ void analogWrite(uint8_t pin, int value) {
+    // Therefore, it will adjust Arduino MAX Duty to be the LEDC MAx Duty
+    uint32_t max_duty = (1 << bus->channel_resolution) - 1;
+    if (value < 0 || value > max_duty) {
+      log_e("Duty is out of range. Valid duty range for pin d is 0 to %d", pin, max_duty);
```
```suggestion
      log_e("Duty is out of range. Valid duty range for pin %d is 0 to %d", pin, max_duty);
```

SuGlider committed:
fix(ledc): missing correct log message format

SuGlider committed:
Merge branch 'master' into ledc_max_resolution_review

SuGlider committed:
Merge branch 'master' into ledc_max_resolution_review

P-R-O-C-H-Y commented:
@lucasssvaz I have changed the milestone to 4.0.0. Also discussed to update the examples and add a note to docs. Tehn we will also add this to migration guide from 3.x to 4.0

SuGlider committed:
feat(ledc): Fix target duty value for LEDC to 4096 (maximum)

SuGlider committed:
feat(ledc): Fix target duty value for LEDC to 4096 (maximum)

SuGlider committed:
feat(ledc): Change LED fade amount and brightness limits

SuGlider committed:
feat(ledc): Adjust LED fade parameters for new maximum values

SuGlider committed:
feat(ledc): Change ledc full brightness to 256

SuGlider committed:
fix(ledc): revert back Hue range to 0..255

SuGlider committed:
feat(ledc): Specify maximum duty value for LEDC functions

Added maximum value constraints for duty parameters in LEDC functions.

SuGlider commented:
> @lucasssvaz I have changed the milestone to 4.0.0. Also discussed to update the examples and add a note to docs. Tehn we will also add this to migration guide from 3.x to 4.0

@P-R-O-C-H-Y - I have modified the examples and documentation. Please review it and check if you agree to it.

lucasssvaz committed:
Merge branch 'master' into ledc_max_resolution_review

lucasssvaz commented:
Updated target branch for IDF 6
